faiss_vector_posting.go

```diff
@@ -266,7 +267,44 @@ func (vpl *VecPostingsIterator) BytesWritten() uint64 {
 	return 0
 }
 
-func (sb *SegmentBase) SimilarVectors(field string, qVector []float32, k int64, except *roaring.Bitmap) (segment.VecPostingsList, error) {
+func (sb *SegmentBase) ReadVectorIndex(field string) (index.VectorIndex, error) {
+	var vecIndex *faiss.IndexImpl
+
+	fieldIDPlus1 := sb.fieldsMap[field]
+	if fieldIDPlus1 > 0 {
+		vectorSection := sb.fieldsSectionsMap[fieldIDPlus1-1][SectionFaissVectorIndex]
+		// check if the field has a vector section in the segment.
+		if vectorSection > 0 {
+			pos := int(vectorSection)
+
+			// loading doc values - adhering to the sections format. never
+			// valid values for vector section
+			_, n := binary.Uvarint(sb.mem[pos : pos+binary.MaxVarintLen64])
+			pos += n
+
+			_, n = binary.Uvarint(sb.mem[pos : pos+binary.MaxVarintLen64])
+			pos += n
+
+			// todo: not a good idea to cache the vector index perhaps, since it could be quite huge.
+			indexSize, n := binary.Uvarint(sb.mem[pos : pos+binary.MaxVarintLen64])
+			pos += n
+			indexBytes := sb.mem[pos : pos+int(indexSize)]
+			pos += int(indexSize)
+
+			var err error
+			vecIndex, err = faiss.ReadIndexFromBuffer(indexBytes, faiss.IOFlagReadOnly)
+			if err != nil {
+				return vecIndex, err
+			}
+		}
+	}
+
+	return vecIndex, nil
+}
+
+func (sb *SegmentBase) SearchSimilarVectors(vecIndex index.VectorIndex, field string, qVector []float32, k int64,
```

Comment:
This API definition doesn't feel good - where it's a member of `SegmentBase` but will only read content from vecIndex which was obtained previously from `sb`. But i understand why you need it this way - so given that it's an exported API, let's add a check here `if vecIndex == nil` - in which case we call GetVectorIndex(..) above.

Also, more importantly - it seems we're repeating a bunch of things we already do in GetVectorIndex above below - which needs some consolidation?

Reply:
I forgot to bring this up during the dev call but I pushed a fix for making this a common func(point 2).
Pls let me know what you think.